File: wrappers/csharp/LibRangeMap/CategoricalRangeMapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class CategoricalRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "categorical_range";

    private readonly object[] _vocabulary;
    private readonly Dictionary<string, int> _vocabularyIndex;

    public double OutputMin { get; }
    public double OutputMax { get; }
    public string? Name { get; }

    public CategoricalRangeMapper(object[] vocabulary, double outputMin = -1.0, double outputMax = 1.0, string? name = null)
    {
        if (vocabulary is null)
        {
            throw new ArgumentNullException(nameof(vocabulary), "vocabulary payload must not be null.");
        }
        if (vocabulary.Length == 0)
        {
            throw new ArgumentException("vocabulary must not be empty.", nameof(vocabulary));
        }

        ValidateFinite(nameof(outputMin), outputMin);
        ValidateFinite(nameof(outputMax), outputMax);
        if (outputMax <= outputMin)
        {
            throw new ArgumentException("output_range must be ordered from lower value to higher value.");
        }
        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }

        _vocabulary = vocabulary.Select(NormalizeToken).ToArray();
        OutputMin = outputMin;
        OutputMax = outputMax;
        Name = name;
        _vocabularyIndex = BuildVocabularyIndex(_vocabulary);
    }

    public double MapValue(object? value)
    {
        if (!IsSupportedToken(value))
        {
            throw new ArgumentException("value must be a string, finite number, boolean, character, or null.");
        }

        var key = TokenKey(value);
        if (!_vocabularyIndex.TryGetValue(key, out var index))
        {
            throw new KeyNotFoundException($"unknown token: {ValueToString(value)}");
        }

        if (_vocabulary.Length == 1)
        {
            return OutputMin;
        }

        return OutputMin + (index / (_vocabulary.Length - 1.0)) * (OutputMax - OutputMin);
    }

    public double Map(object? value) => MapValue(value);

    public string ToJson() => JsonSerializer.Serialize(ToSpec(), JsonOptions);

    public static CategoricalRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<CategoricalMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static CategoricalRangeMapper FromSpec(CategoricalMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }
        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }
        if (spec.Vocabulary is null)
        {
            throw new ArgumentException("vocabulary is required.", nameof(spec));
        }
        if (spec.Vocabulary.Length == 0)
        {
            throw new ArgumentException("vocabulary must not be empty.", nameof(spec));
        }
        if (spec.OutputRange is null || spec.OutputRange.Length != 2)
        {
            throw new ArgumentException("output_range must contain exactly two values.", nameof(spec));
        }

        return new CategoricalRangeMapper(spec.Vocabulary.Select(DecodeToken).ToArray(), spec.OutputRange[0], spec.OutputRange[1], spec.Name);
    }

    private CategoricalMapperSpec ToSpec()
    {
        return new CategoricalMapperSpec
        {
            SpecVersion = SpecVersion,
            MapperType = SpecType,
            Vocabulary = _vocabulary.Select(EncodeToken).ToArray(),
            OutputRange = new[] { OutputMin, OutputMax },
            Name = Name
        };
    }

    private static string? EncodeToken(object? token)
    {
        return token switch
        {
            null => null,
            string text => "str:" + text,
            char ch => "char:" + (int)ch,
            bool boolValue => "bool:" + (boolValue ? "true" : "false"),
            byte v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            sbyte v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            short v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            ushort v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            int v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            uint v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            long v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            ulong v => "num:" + v.ToString(System.Globalization.CultureInfo.InvariantCulture),
            float v => "num:" + FormatNumber(v),
            double v => "num:" + FormatNumber(v),
            decimal v => "num:" + FormatNumber(v),
            _ => throw new ArgumentException($"unsupported vocabulary token type: {token?.GetType().FullName}")
        };
    }

    private static object? DecodeToken(string? encoded)
    {
        if (encoded is null)
        {
            return null;
        }

        var separator = encoded.IndexOf(':');
        if (separator < 0)
        {
            return encoded;
        }

        var kind = encoded[..separator];
        var raw = encoded[(separator + 1)..];
        return kind switch
        {
            "str" => raw,
            "char" => (char)int.Parse(raw, System.Globalization.CultureInfo.InvariantCulture),
            "bool" => bool.Parse(raw),
            "num" => ParseNumber(raw),
            _ => throw new ArgumentException($"unsupported token encoding: {encoded}")
        };
    }

    private static object NormalizeToken(object token)
    {
        if (!IsSupportedToken(token))
        {
            throw new ArgumentException($"unsupported vocabulary token type: {token.GetType().FullName}");
        }

        return token switch
        {
            string text => text,
            char ch => ch,
            bool value => value,
            byte or sbyte or short or ushort or int or uint or long or ulong => token,
            float floatValue => floatValue,
            double doubleValue => doubleValue,
            decimal decimalValue => decimalValue,
            null => null!,
            _ => throw new ArgumentException($"unsupported vocabulary token type: {token.GetType().FullName}")
        };
    }

    private static bool IsSupportedToken(object? value)
    {
        return value switch
        {
            null => true,
            string => true,
            bool => true,
            char => true,
            byte or sbyte or short or ushort or int or uint or long or ulong => true,
            float v => float.IsFinite(v),
            double v => double.IsFinite(v),
            decimal => true,
            _ => false
        };
    }

    private static string TokenKey(object? value)
    {
        if (value is null)
        {
            return "null";
        }

        return value switch
        {
            string text => "str:" + text,
            bool boolValue => "bool:" + boolValue,
            char ch => "char:" + (int)ch,
            byte v => "num:" + v,
            sbyte v => "num:" + v,
            short v => "num:" + v,
            ushort v => "num:" + v,
            int v => "num:" + v,
            uint v => "num:" + v,
            long v => "num:" + v,
            ulong v => "num:" + v,
            float v => float.IsFinite(v) ? "num:" + FormatNumber(v) : throw new ArgumentException("vocabulary tokens must be finite numbers."),
            double v => double.IsFinite(v) ? "num:" + FormatNumber(v) : throw new ArgumentException("vocabulary tokens must be finite numbers."),
            decimal v => "num:" + FormatNumber(v),
            _ => throw new ArgumentException("value must be a string, finite number, boolean, character, or null.")
        };
    }

    private static string FormatNumber(double value)
    {
        return value % 1 == 0 ? ((long)value).ToString() : value.ToString("R");
    }

    private static string FormatNumber(float value)
    {
        return value % 1 == 0 ? ((long)value).ToString() : value.ToString("R");
    }

    private static string FormatNumber(decimal value)
    {
        return decimal.Truncate(value) == value ? decimal.ToInt64(value).ToString() : value.ToString(System.Globalization.CultureInfo.InvariantCulture);
    }

    private static object ParseNumber(string raw)
    {
        var number = double.Parse(raw, System.Globalization.CultureInfo.InvariantCulture);
        if (!double.IsFinite(number))
        {
            throw new ArgumentException("vocabulary tokens must be finite numbers.");
        }

        var asLong = (long)number;
        if (number == asLong)
        {
            return asLong;
        }

        return number;
    }

    private static Dictionary<string, int> BuildVocabularyIndex(object[] vocabulary)
    {
        var index = new Dictionary<string, int>(vocabulary.Length, StringComparer.Ordinal);
        for (var i = 0; i < vocabulary.Length; i++)
        {
            var token = vocabulary[i];
            var key = TokenKey(token);
            if (index.ContainsKey(key))
            {
                throw new ArgumentException($"vocabulary contains duplicate token: {ValueToString(token)}");
            }
            index[key] = i;
        }

        return index;
    }

    private static string ValueToString(object? value)
    {
        return value switch
        {
            null => "null",
            string text => $"\"{text}\"",
            char ch => $"'{ch}'",
            _ => value.ToString() ?? string.Empty
        };
    }

    private static void ValidateFinite(string label, double value)
    {
        if (!double.IsFinite(value))
        {
            throw new ArgumentException($"{label} must be finite.", label);
        }
    }
}

public sealed class CategoricalMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "categorical_range";

    [JsonPropertyName("vocabulary")]
    public string?[]? Vocabulary { get; init; }

    [JsonPropertyName("output_range")]
    public double[]? OutputRange { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}

File: wrappers/vb/LibRangeMap/CategoricalRangeMapper.vb
Imports System.Collections.Generic
Imports System.Text.Json

Public Class CategoricalRangeMapper
    Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
        .PropertyNamingPolicy = Nothing,
        .WriteIndented = False
    }

    Private ReadOnly _vocabulary As String()
    Private ReadOnly _outputMin As Double
    Private ReadOnly _outputMax As Double

    Public Sub New(vocabulary As String(), Optional outputMin As Double = -1.0, Optional outputMax As Double = 1.0)
        If vocabulary Is Nothing Then
            Throw New ArgumentNullException(NameOf(vocabulary))
        End If
        If vocabulary.Length = 0 Then
            Throw New ArgumentException("vocabulary must not be empty.", NameOf(vocabulary))
        End If
        If Double.IsNaN(outputMin) OrElse Double.IsInfinity(outputMin) OrElse Double.IsNaN(outputMax) OrElse Double.IsInfinity(outputMax) Then
            Throw New ArgumentException("output range endpoints must be finite.")
        End If
        If outputMin >= outputMax Then
            Throw New ArgumentException("outputMin must be less than outputMax.")
        End If

        Dim seen As New HashSet(Of String)(StringComparer.Ordinal)
        For Each token In vocabulary
            If token Is Nothing OrElse token.Length = 0 Then
                Throw New ArgumentException("vocabulary tokens must be non-empty strings.", NameOf(vocabulary))
            End If
            If Not seen.Add(token) Then
                Throw New ArgumentException($"duplicate vocabulary token '{token}'.", NameOf(vocabulary))
            End If
        Next

        _vocabulary = DirectCast(vocabulary.Clone(), String())
        _outputMin = outputMin
        _outputMax = outputMax
    End Sub

    Public Function MapValue(value As String) As Double
        If value Is Nothing Then
            Throw New ArgumentNullException(NameOf(value))
        End If

        Dim index As Integer = Array.IndexOf(_vocabulary, value)
        If index < 0 Then
            Throw New ArgumentException($"unknown categorical token '{value}'.", NameOf(value))
        End If

        If _vocabulary.Length = 1 Then
            Return (_outputMin + _outputMax) / 2.0
        End If

        Return _outputMin + ((CDbl(index) / CDbl(_vocabulary.Length - 1)) * (_outputMax - _outputMin))
    End Function

    Public Function Map(value As String) As Double
        Return MapValue(value)
    End Function

    Public Function Spec() As CategoricalMapperSpec
        Return New CategoricalMapperSpec With {
            .SpecVersion = CurrentSpecVersion,
            .MapperType = "categorical_range",
            .Vocabulary = DirectCast(_vocabulary.Clone(), String()),
            .OutputRange = New Double() {_outputMin, _outputMax}
        }
    End Function

    Public Function ToJson() As String
        Return JsonSerializer.Serialize(Spec(), JsonOptions)
    End Function

    Public Shared Function FromJson(json As String) As CategoricalRangeMapper
        Dim spec = JsonSerializer.Deserialize(Of CategoricalMapperSpec)(json, JsonOptions)
        If spec Is Nothing Then
            Throw New InvalidOperationException("mapper spec JSON produced null.")
        End If
        Return FromSpec(spec)
    End Function

    Public Shared Function FromSpec(spec As CategoricalMapperSpec) As CategoricalRangeMapper
        If spec.SpecVersion <> CurrentSpecVersion Then
            Throw New ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", NameOf(spec))
        End If
        If spec.MapperType <> "categorical_range" Then
            Throw New ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", NameOf(spec))
        End If
        If spec.Vocabulary Is Nothing OrElse spec.Vocabulary.Length = 0 Then
            Throw New ArgumentException("vocabulary must contain at least one token.", NameOf(spec))
        End If
        If spec.OutputRange Is Nothing OrElse spec.OutputRange.Length <> 2 Then
            Throw New ArgumentException("output_range must contain exactly two values.", NameOf(spec))
        End If

        Return New CategoricalRangeMapper(spec.Vocabulary, spec.OutputRange(0), spec.OutputRange(1))
    End Function
End Class
